ping-with-dsl/testplans.ts

```diff
@@ -0,0 +1,66 @@
+import { TestPlans } from "dsl/src/dsl"
+import { run } from "dsl/src/testground-runner"
+import goV0230 from "./go/v0.23.0/image.json"
+import rustV0490 from "./rust/image.0.49.0.json"
+
+type PingParams = {
+    max_latency_ms: number,
+    iterations: number,
+}
+
+const testplans: TestPlans<PingParams> = {
+    testPlans: [
+        {
+            name: "Go v0.23",
+            instances: [
+                {
+                    name: "Go v0.23 x Go v0.23",
+                    containerImageID: goV0230.imageID,
+                    copies: 2,
+                    runtimeEnv: {
+                        "max_latency_ms": 500,
+                        "iterations": 5,
+                    }
+                },
+            ]
+        },
+        {
+            name: "Rust v0.49.0",
+            instances: [
+                {
+                    name: "Rust v0.49.0 x Rust v0.49.0",
+                    containerImageID: rustV0490.imageID,
+                    copies: 2,
+                    runtimeEnv: {
+                        "max_latency_ms": 500,
+                        "iterations": 5,
+                    }
+                }
+            ]
+
+        },
+        {
+            name: "Go v0.23 x Rust v0.49.0",
+            instances: [
+                {
+                    name: "Go v0.23",
+                    containerImageID: goV0230.imageID,
+                    runtimeEnv: {
+                        "max_latency_ms": 500,
+                        "iterations": 5,
+                    }
+                },
+                {
+                    name: "Rust v0.49.0",
+                    containerImageID: rustV0490.imageID,
+                    runtimeEnv: {
+                        "max_latency_ms": 500,
+                        "iterations": 5,
+                    }
+                }
+            ]
+        }
+    ]
+}
```

Comment:
We're creating a DSL / configuration abstraction above Compositions:

This PR does two things: move the build step out of Testground AND introduce a new configuration format.

The new configuration looks MUCH simpler because it IS: we removed the essential complexity of the building configuration.

With a regular composition (no change to Testground required), this is what the same workflow would look like:
https://gist.github.com/laurentsenta/bde72a8ad2e8cfdcecfb3add75338a44

Turned into yaml which is more familiar than toml:
https://gist.github.com/laurentsenta/f85b2249b1598586a01071b61862e68a

I would recommend not going for a complete "DSL" abstraction yet.

Reply:
> With a regular composition (no change to Testground required), this is what the same workflow would look like:

I know this is an example, but this doesn't work today. You still need a plan and a test case. These are vestiges that I think unnecessarily complicated.

If we want to iterate to something closer to this DSL (minimal config) and change this DSL later, that's fine by me. But I don't want to block this work on first changing testground.

Reply:
A couple other things when comparing this to a regular composition file:
- The DSL leverages typescript to typecheck the runtime params
- The composition file has the idea of "groups" and "runs" as separate things. This is vestige of having testground build things. But really we only need "runs" here. 

Here's both configs in JSON: https://gist.github.com/MarcoPolo/e4488ce57c58469298918af896ded192. This is subjective, but I find the DSL version simpler.

Reply:
> I know this is an example, but this doesn't work today. You still need a plan and a test case. These are vestiges that I think unnecessarily complicated.

The example would work, except for a typo with artifacts maybe.
I'm confused because we could do:

```
params -- (generator) --> composition
```

instead of:

```
params -- (generator) --> "dsl" -- (transformer) --> composition
```

- typing is supported just as well, type the boundaries of your generator, you could even have fun with generics there,
- the main change in the DSL is that we're going to duplicate every group, so code generation will be the default, no one is going to manually edit this file.
- there's nothing to block for in Testground.

It seems more efficient. Let's have a chat async on why we can't do this and why we call this a DSL instead of a new configuration format, Feel free to resolve this whenever you want.

Reply:
> The example would work, except for a typo with artifacts maybe.

At the very least you currently need a `manifest.toml` that defines a the plan name, defaults, builders, runners, and testcases.

> the main change in the DSL is that we're going to duplicate every group, so code generation will be the default, no one is going to manually edit this file.

I don't follow, sorry.

> there's nothing to block for in Testground.

What's blocking now?

>  why we call this a DSL instead of a new configuration format,

A rose by any other name. Happy to call this a new configuration format.